### ISSUE TITLE
Check if object is NaN or is Infinity while converting to Phi object

### ISSUE THREAD

**Issue by maxonfjvipon** [objectionary/eo#1253]
In some cases we can get `NaN` or `Infinities` as a result of object dataization.   
For example:  
- `0.0.div 0.0` => `nan`, 
- `1.0.div 0.0` => `positive-infinity`, 
- `(number nan).ln` => `nan` (natural logarithm, in progress)
- etc...

NaN and Ifinities are double, so I suggest to do something like this:
```java
private static Phi toPhi(final Object obj) {
    ...
    } else if (obj instanceof Double) {
        if (((Double) obj).isNaN()) {
            phi = new EOnan(Phi.Φ);
        } else {
            phi = new EOfloat(Phi.Φ);
        }
    }
    ....
}
```

And also add checking infinities here.  
In that case we won't get an error when we do for example `0.0.div 0.0`, we'll get `nan` object and will be able to work with it

**Comment by yegor256** [comment]
@maxonfjvipon but we don't have `EOnan` in here

**Comment by maxonfjvipon** [comment]
@yegor256 will not it be created on runtime?

**Comment by yegor256** [comment]
@maxonfjvipon it belongs to eo-math package. We can't use objects from this package inside eo-runtime. Well, technically we can, but this would be violation of "layered architecture" principle.

**Comment by Graur** [comment]
@maxonfjvipon @yegor256  We already talk about it here: https://github.com/objectionary/eo/issues/774#issuecomment-1164312509

**Comment by maxonfjvipon** [comment]
@yegor256 yeah, I remember, but it's more complicated)  
Ok, I try to figure it out, but I think eventually it must be done somehow. `0.0.div 0.0` has not to throw an exception. It should return `nan` object.

**Comment by maxonfjvipon** [comment]
@yegor256 I did a little research and here's what I think. 
Anyway we should move `nan` and `infinities` to `eo-runtime`. I believe it's the place where they belong to. Even in Java `NaN` belong to `Double` class. In other cases we won't be able to solve this problem.

After we move them we can solve it in two ways.
1. Easy. As I described above - add extra conditions to `toPhi`. There'll be no cross layering anymore because `nan`,`infinities` and `float` are in the same package.
2. Not easy and maybe not even worth to be discussed. As I suggested [here](https://github.com/objectionary/eo/issues/774): try to add some new EO syntax.

**Comment by yegor256** [comment]
@maxonfjvipon I have another proposal. How about we make sure `div` doesn't throw an exception but returns `float`, which is `nan` (byte wise)?

**Comment by maxonfjvipon** [comment]
@yegor256 so the result of `0.0.div 0.0` will be `float` object, but with bytes of `nan`. 
```
0.0.div 0.0 > x # EOfloat
QQ.math.nan > n # EOnan
```

Here we should make sure that `x` and `n` have the same amount of inner objects with the same behavior

**Comment by maxonfjvipon** [comment]
@yegor256 It seems some objects like `float.lt|lte|gte` may not work correctly. Even `float.eq`

```
(0.0.div 0.0).eq nan # it'll be TRUE, but has to be FALSE
```

**Comment by maxonfjvipon** [comment]
@Graur 0.0 / 0.0 = nan, 1.0 / 0.0 = +inf, -1.0 / 0.0 = -inf

**Comment by maxonfjvipon** [comment]
Also we need somehow to check `nan` here (and in some other places) in case we do `4.2.div nan`.

<img width="528" alt="image" src="https://user-images.githubusercontent.com/37025995/191725146-c39cc839-3d59-4924-b6f5-a56a7b55eec2.png">

We could do something like this: make `nan` decorated `0.0.div 0.0`
```
[] > nan
  0.0.div 0.0 > @
```
Then when we do `3.2.div nan`, in Java `nan` will be dataized to `Double.NaN` and 3.2 / NaN would return Double.NaN.

But I see some kind of conflict here: `0.0.div 0.0` should be cast to `EOnan` in Java, and if we make `nan` decorated `0.0.div 0.0` then `nan` decorates itself. 

Also if we could do something like this it could solve the whole problem
```
[] > nan
  "java.lang.Double.NaN" > @
```

**Comment by maxonfjvipon** [comment]
Resume:
1. Move `nan`, `positive-infinity` and `negative-infinity` to `eo-runtime` and place them inside `float` object
2. Add decoratees to these objects so they can be dataized to java primitives
3. Allow division by float zero
4. Add extra checks to `Data.ToPhi.toPhi()`, for example:
```java
if (((Double) obj).isNaN()) {
    phi = new EOfloat$EOnan(Phi.Φ);
}
```

I believe I didn't miss anything. 
@Graur please assign me. Can you set a 60 minutes budget?

**Comment by Graur** [comment]
@maxonfjvipon Sure, but it's always better to split this task to several puzzles by 30 min each. Please, use [pdd](https://www.yegor256.com/2017/04/05/pdd-in-action.html) for such cases. 
I suggest at least 4 pdd puzzles with 30 min budget each.

**Comment by maxonfjvipon** [comment]
@Graur Ok, sounds great!

**Comment by 0pdd** [comment]
@maxonfjvipon 3 puzzles [#1800](https://github.com/objectionary/eo/issues/1800), [#1801](https://github.com/objectionary/eo/issues/1801), [#1802](https://github.com/objectionary/eo/issues/1802) are still not solved.

**Comment by 0pdd** [comment]
@maxonfjvipon all 3 puzzles are solved here: [#1800](https://github.com/objectionary/eo/issues/1800), [#1801](https://github.com/objectionary/eo/issues/1801), [#1802](https://github.com/objectionary/eo/issues/1802).
